### PR TITLE
[OP-1966] MongoDB terraform provider

### PR DESCRIPTION
**Pull request by jilleJr**
OP-1966

Based on template repo: https://github.com/hashicorp/terraform-provider-scaffolding-framework

Followed guide at: https://developer.hashicorp.com/terraform/tutorials/providers-plugin-framework/providers-plugin-framework-provider

Overarching changes:

- Removed example code
- Added markdown & license linting
- Added REUSE-compliant license annotations
- Implemented:
  - Provider `mongodb_user`
  - Data source `data.mongodb_users`

### CODE REVIEW

File: internal/mongodb/mongodb.go [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+		return User{}, err
+	}
+	user, err := c.GetDBUser(ctx, dbName, newUser.User)
+	if err != nil {
+		return User{}, fmt.Errorf("get created user: %w", err)
+	}
+	return user, nil
+}
+
+func (c *Client) createDBUser(ctx context.Context, dbName string, newUser NewUser) error {
```
Comment (jadolg) [inline internal/mongodb/mongodb.go:239]: This pattern is confusing to me. Internally, I wouldn't know at first which function to use. What's the purpose of separation here? Do they have the same responsibility?

File: LICENSES/CC0-1.0.txt [hunk 2]
```diff
@@ -0,0 +1,121 @@
+Creative Commons Legal Code
```
Comment (jadolg) [inline LICENSES/CC0-1.0.txt:1]: why all these many licenses?

File: examples/provider-install-verification/main.tf [hunk 3]
```diff
@@ -0,0 +1,16 @@
+
+terraform {
+  required_providers {
+    mongodb = {
+      source = "hashicorp.com/edu/mongodb-driver"
```
Comment (jadolg) [inline examples/provider-install-verification/main.tf:5]: nit: It'd be nice to have a more personalized URL

File: examples/provider-install-verification/main.tf [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  }
+}
+
+provider "mongodb" {
+  host     = "http://localhost:19090"
+  username = "education"
+  password = "test123"
+}
+
+data "mongodb_coffees" "example" {}
```
Comment (jadolg) [inline examples/provider-install-verification/main.tf:16]: coffees?

File: examples/provider-install-verification/main.tf (hunk 3 above)
Comment (jilleJr) [reply examples/provider-install-verification/main.tf:5]: Ah yes I should update this one. This naming is from the tutorial I followed.

The name should be something like `registry.terraform.io/riskident/mongodb-driver`, depending on the username we can register at registry.terraform.io.

Will change

File: internal/mongodb/mongodb.go (hunk 1 above)
Comment (jilleJr) [reply internal/mongodb/mongodb.go:239]: The unpublished/lowercase one is only creating the user without returning it, as the `createUser` MongoDB command doesn't return the created user. I wanted to have "just call the `createUser` command" as a separate function, to encapsulate the context timeout and other local variables.

I don't know how to name it to signify this.

Any suggestions? I can only come up with weird names like:

- `createButDontGetDBUser`
- `onlyCreateDBUser`
- `runCreateUserCommand`

Maybe we could rename the published method? Like:

- `CreateAndGetDBUser`
- Not having both in same method, force caller to use both `CreateDBUser` and then `GetDBUser` explicitly

File: examples/provider-install-verification/main.tf (hunk 4 above)
Comment (jilleJr) [reply examples/provider-install-verification/main.tf:16]: Oops, this is no longer needed.

File: internal/mongodb/mongodb.go [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+func New(uri string, cred Credentials) *Client {
+	return &Client{
+		uri:         uri,
+		credentials: cred,
+	}
+}
+
+func (c *Client) connect(ctx context.Context) (client *mongo.Client, connectErr error) {
+	c.connectOnce.Do(func() {
+		ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
```
Comment (jadolg) [inline internal/mongodb/mongodb.go:48]: nit: Maybe this timeout should be configurable?

File: internal/mongodb/mongodb.go (hunk 1 above)
Comment (jilleJr) [reply internal/mongodb/mongodb.go:239]: Any suggestions on the name? Personally not super fond of `CreateAndGetDBUser`

File: internal/mongodb/mongodb.go (hunk 5 above)
Comment (jilleJr) [reply internal/mongodb/mongodb.go]: Removed timeouts from `mongodb` package and instead provided them from the `provider` package. Let's users define timeouts "the de-facto terraform way"

https://developer.hashicorp.com/terraform/plugin/framework/resources/timeouts

File: internal/mongodb/mongodb.go (hunk 1 above)
Comment (jilleJr) [reply internal/mongodb/mongodb.go]: Updated in bc3e325, wdyt?

Comment (jadolg) [reply internal/mongodb/mongodb.go]: Personally, I liked more the CreateAndGetDBUser option since it describes exactly what the function does.

**Review by jadolg — APPROVED** [review]

**Comment by jadolg** [comment]
I'm still not very convinced about the [naming](https://github.com/RiskIdent/terraform-provider-mongodb-driver/pull/2#discussion_r1271841244), but that's not a blocker

File: internal/mongodb/mongodb.go (hunk 1 above)
Comment (jilleJr) [reply internal/mongodb/mongodb.go]: I don't like having the term "and" inside a method or variable name. Maybe something like "NewDBUser" would be more appropriate? As "New" does suggest something is created, and it returns the created object.

Though the term "New" is conventionally just a constructor in Go, so you'd expect a "New-" function to not have side effects... Hmm! :thinking:

Considering you said the naming was not a blocker, then I'm resolving

**Pull request merged**